internal/chain/engine.go

```diff
@@ -145,6 +146,34 @@ func ReadPublicKeysFromLastBlock(bc engine.ChainReader, header *block.Header) ([
 	return GetPublicKeys(bc, parentHeader, false)
 }
 
+// VerifyShardState implements Engine, checking the shardstate is valid at epoch transition
+func (e *engineImpl) VerifyShardState(bc engine.ChainReader, beacon engine.ChainReader, header *block.Header) error {
+	headerShardState, _ := header.GetShardState()
```

Comment:
Just read the raw bytes directly. This first decode to shard.State. And you later encode it again... no need.

Reply:
This is when there are issue to debug. will change back